File: programs/dippies_index_protocol/src/instructions/create_note.rs
use anchor_lang::prelude::*;

use crate::constants::{
    FOREST_SEED, MAX_DESCRIPTION_LENGTH, MAX_URI_LENGTH, NODE_SEED, NOTE_SEED, TREE_SEED,
};
use crate::errors::DipErrors;
use crate::state::{Forest, Node, Note, Tree};

pub fn create_note(
    ctx: Context<CreateNote>,
    id: Pubkey,
    title: String,
    website: String,
    image: String,
    description: String,
) -> Result<()> {
    msg!("Creating a note");

    let note = &mut ctx.accounts.note;
    note.id = id;
    note.tree = ctx.accounts.tree.key();
    note.parent = ctx.accounts.node.key();
    note.stake = 0;
    note.title = title;
    note.website = website;
    note.image = image;
    note.description = description;
    note.tags = ctx.accounts.node.tags.clone();

    Ok(())
}

#[derive(Accounts)]
#[instruction(id: Pubkey, website: String, image: String, description: String)]
pub struct CreateNote<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,

    /// The global forest
    #[account(
        seeds = [
            FOREST_SEED.as_bytes(),
            &forest.id.to_bytes(),
        ],
        bump,
    )]
    pub forest: Account<'info, Forest>,

    /// The tree
    #[account(
        seeds = [
            TREE_SEED.as_bytes(),
            &forest.key().to_bytes(),
            &tree.title.as_ref(),
        ],
        bump,
    )]
    pub tree: Account<'info, Tree>,

    /// The node to attach to
    #[account(
        seeds = [
            NODE_SEED.as_bytes(),
            &tree.key().to_bytes(),
            &node.parent.key().to_bytes(),
            &node.tags.last().unwrap().as_ref(),
        ],
        bump,
        constraint = node.children.len() == 0 @ DipErrors::InvalidNode
    )]
    pub node: Account<'info, Node>,

    /// The new note
    #[account(
        init,
        payer = signer,
        space = Note::LEN,
        seeds = [
            NOTE_SEED.as_bytes(),
            &tree.key().to_bytes(),
            &id.to_bytes()
        ],
        bump,
        constraint = website.len() <= MAX_URI_LENGTH @ DipErrors::StringTooLong,
        constraint = website.len() <= MAX_URI_LENGTH @ DipErrors::StringTooLong,
        constraint = image.len() <= MAX_URI_LENGTH @ DipErrors::StringTooLong,
        constraint = description.len() <= MAX_DESCRIPTION_LENGTH @ DipErrors::StringTooLong,
    )]
    pub note: Account<'info, Note>,

    /// Common Solana programs
    pub system_program: Program<'info, System>,
    pub rent: Sysvar<'info, Rent>,
}

File: ts/instructions/createNote.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface CreateNoteArgs {
  id: PublicKey
  website: string
  image: string
  description: string
}

export interface CreateNoteAccounts {
  signer: PublicKey
  /** The global forest */
  forest: PublicKey
  /** The tree */
  tree: PublicKey
  /** The node to attach to */
  node: PublicKey
  /** The new note */
  note: PublicKey
  /** Common Solana programs */
  systemProgram: PublicKey
  rent: PublicKey
}

export const layout = borsh.struct([
  borsh.publicKey("id"),
  borsh.str("website"),
  borsh.str("image"),
  borsh.str("description"),
])

export function createNote(args: CreateNoteArgs, accounts: CreateNoteAccounts) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.signer, isSigner: true, isWritable: true },
    { pubkey: accounts.forest, isSigner: false, isWritable: false },
    { pubkey: accounts.tree, isSigner: false, isWritable: false },
    { pubkey: accounts.node, isSigner: false, isWritable: false },
    { pubkey: accounts.note, isSigner: false, isWritable: true },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
    { pubkey: accounts.rent, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([103, 2, 208, 242, 86, 156, 151, 107])
  const buffer = Buffer.alloc(1000)
  const len = layout.encode(
    {
      id: args.id,
      website: args.website,
      image: args.image,
      description: args.description,
    },
    buffer
  )
  const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len)
  const ix = new TransactionInstruction({ keys, programId: PROGRAM_ID, data })
  return ix
}
